### PR TITLE
Print template paths

### PR DESCRIPTION
**Pull request by balazsbajorics**
**Problem:**
Only UIDs are printed to the DOM, and we use the validPaths array to backsolve which template path might belong to which UID. There's a lot of edge cases that need dancing around. In the meantime, we realized that the old limitations that prevented printing the full TemplatePath don't exist anymore. 

**Fix:**
Use template paths in the DOM

**Commit Details:**
- Print the full template paths to the DOM, 
- Apply the same monkey patching to the printed template paths that we apply to UIDs
- use those in the dom-walker instead of backsolving based on uids
- make templatePath non-optional in the element-renderer functions (it didn't need to be optional anymore anyways)
- update a bunch of test snapshots

### CODE REVIEW

File: editor/src/components/canvas/ui-jsx-canvas.tsx [hunk 1]
```diff
@@ -88,6 +88,9 @@ import { JSX_CANVAS_LOOKUP_FUNCTION_NAME } from '../../core/workers/parser-print
 import { getParseSuccessOrTransientForFilePath } from './ui-jsx-canvas-renderer/ui-jsx-canvas-top-level-elements'
 import { ProjectContentTreeRoot, getContentsTreeFileFromString } from '../assets'
 import { createExecutionScope } from './ui-jsx-canvas-renderer/ui-jsx-canvas-execution-scope'
+import { applyUIDMonkeyPatch } from '../../utils/canvas-react-utils'
+
+applyUIDMonkeyPatch()
```
Comment (enidemi) [inline editor/src/components/canvas/ui-jsx-canvas.tsx:93]: why is this changed?

**Review by enidemi — APPROVED** [review]
I like it!

File: editor/src/components/canvas/__snapshots__/dom-walker.spec.browser.tsx.snap [hunk 2]
```diff
@@ -262,6 +263,7 @@ Object {
       "y": 0,
     },
     "props": Object {
+      "data-paths": "utopia-storyboard-uid/scene-aaa:aaa",
```
Comment (Rheeseyb) [inline editor/src/components/canvas/__snapshots__/dom-walker.spec.browser.tsx.snap:266]: Hmmm a lot of these seem wrong. Where they should have two paths (according to the `data-uid` prop below) they only have one.

File: editor/src/components/canvas/ui-jsx-canvas.spec.tsx [hunk 3]
```diff
@@ -1234,9 +1240,14 @@ export var ${BakedInStoryboardVariableName} = (props) => {
...
             height: 400px;
           \\"
+          data-paths=\\":utopia-storyboard-uid\\"
           data-uid=\\"utopia-storyboard-uid\\"
         >
-          <div id=\\"nasty-div\\" data-uid=\\"2a7~~~1 150~~~2 2f5~~~1 aaa scene-aaa\\">
+          <div
+            id=\\"nasty-div\\"
+            data-uid=\\"2a7~~~1 150~~~2 2f5~~~1 aaa scene-aaa\\"
+            data-paths=\\":150~~~2/2a7~~~1 :150~~~2 :2f5~~~1 utopia-storyboard-uid/scene-aaa:aaa\\"
```
Comment (Rheeseyb) [inline editor/src/components/canvas/ui-jsx-canvas.spec.tsx:1249]: Hmmm yeah something is definitely wrong here

File: editor/src/components/canvas/__snapshots__/dom-walker.spec.browser.tsx.snap (hunk 2 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/__snapshots__/dom-walker.spec.browser.tsx.snap]: I pushed a fix!

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap [hunk 4]
```diff
@@ -21572,21 +22062,30 @@ exports[`UiJsxCanvas render renders fragments correctly 1`] = `
...
         style=\\"width: 100px; height: 100px;\\"
         data-uid=\\"bbb\\"
+        data-paths=\\"eee/fff:aaa/25d80840-fbda-4984-be3b-43838cb78079\\"
       >
         Hello
         <div
           data-label=\\"some-other-div\\"
           style=\\"width: 100px; height: 100px;\\"
           data-uid=\\"ccc\\"
+          data-paths=\\"eee/fff:aaa/25d80840-fbda-4984-be3b-43838cb78079/8e84c2c0-99bd-40dc-bc63-4d8230425414\\"
```
Comment (balazsbajorics) [inline editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:22079]: hmmmmmmmmm, what turns `ccc` into `25d80840-fbda-4984-be3b-43838cb78079` ? 😂 

at least this uid is truly random so this broke the test, helping me find it...

File: editor/src/components/canvas/__snapshots__/dom-walker.spec.browser.tsx.snap [hunk 5]
```diff
@@ -38,6 +38,7 @@ Object {
       "y": 0,
     },
     "props": Object {
+      "data-paths": ":utopia-storyboard-uid",
```
Comment (Rheeseyb) [inline editor/src/components/canvas/__snapshots__/dom-walker.spec.browser.tsx.snap:41]: I can't wait until we can remove the leading `:`

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap [hunk 6]
```diff
@@ -4226,11 +4347,25 @@ exports[`UiJsxCanvas render class component is available from arbitrary block in
...
+      <div
+        data-uid=\\"ccc aaa\\"
+        data-paths=\\"utopia-storyboard-uid/scene-aaa:zzz/aaa\\"
+      >
+        Thing
+      </div>
+      <div
+        data-uid=\\"ccc bbb\\"
+        data-paths=\\"utopia-storyboard-uid/scene-aaa:zzz/bbb\\"
+      >
```
Comment (Rheeseyb) [inline editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:4366]: Looks like something has gone wrong here, because both of these elements should have 2 paths - both are missing a path ending in `:ccc`

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap [hunk 7]
```diff
@@ -15598,9 +15954,15 @@ exports[`UiJsxCanvas render refs are handled and triggered correctly in a class
...
+    data-paths=\\":utopia-storyboard-uid\\"
     data-uid=\\"utopia-storyboard-uid\\"
   >
-    <div data-uid=\\"ccc aaa scene-aaa\\">Thing</div>
+    <div
+      data-uid=\\"ccc aaa scene-aaa\\"
+      data-paths=\\"utopia-storyboard-uid/scene-aaa:aaa :utopia-storyboard-uid/scene-aaa\\"
+    >
+      Thing
+    </div>
```
Comment (Rheeseyb) [inline editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:15965]: This is another case that seems to be missing a path

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap [hunk 8]
```diff
@@ -21228,9 +21710,15 @@ exports[`UiJsxCanvas render renders fine with two components that reference each
...
+    data-paths=\\":utopia-storyboard-uid\\"
     data-uid=\\"utopia-storyboard-uid\\"
   >
-    <div data-uid=\\"2a2~~~1 aaa bbb BBB scene\\">great</div>
+    <div
+      data-uid=\\"2a2~~~1 aaa bbb BBB scene\\"
+      data-paths=\\":2a2~~~1 utopia-storyboard-uid/scene:BBB :utopia-storyboard-uid/scene\\"
+    >
+      great
+    </div>
```
Comment (Rheeseyb) [inline editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:21721]: I'm not sure what has happened with the paths here, but something is definitely wrong :grimacing:

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap [hunk 9]
```diff
@@ -21840,7 +22340,7 @@ export var storyboard = (
       },
     },
   },
-  "eee/fff:aaa/ddd": Object {
+  "eee/fff:aaa/bbb": Object {
```
Comment (Rheeseyb) [inline editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:22343]: This is interesting. Something from here onwards has been parsed or walked completely differently :thinking:

File: editor/src/components/canvas/dom-walker.ts [hunk 10]
```diff
@@ -827,16 +796,21 @@ function walkElements(
   }
   if (element instanceof HTMLElement) {
     // Determine the uid of this element if it has one.
-    const uids = getUIDsOnDomELement(element) ?? []
-    const filteredValidPaths = filterValidFocusedPathsForGeneratedElement(uids, validPaths)
+    const paths = getPathsOnDomElement(element)
+    // const filteredValidPaths = filterValidFocusedPathsForGeneratedElement(uids, validPaths)
```
Comment (Rheeseyb) [inline editor/src/components/canvas/dom-walker.ts:800]:
```suggestion
```

File: editor/src/components/canvas/ui-jsx-canvas-renderer/scene-root.tsx [hunk 11]
```diff
@@ -238,6 +242,7 @@ export const SceneRootRenderer = betterReactMemo(
       ...sceneProps.props,
       [UTOPIA_SCENE_PATH]: scenePath,
       [UTOPIA_UIDS_KEY]: sceneUID,
```
Comment (Rheeseyb) [inline editor/src/components/canvas/ui-jsx-canvas-renderer/scene-root.tsx:244]: Can we bin off the uid key with this change?

File: editor/src/components/canvas/ui-jsx-canvas.spec.tsx [hunk 12]
```diff
@@ -1133,9 +1133,15 @@ export var ${BakedInStoryboardVariableName} = (props) => {
             width: 400px;
             height: 400px;
           \\"
+          data-paths=\\":utopia-storyboard-uid\\"
           data-uid=\\"utopia-storyboard-uid\\"
         >
-          <div data-uid=\\"ccc aaa scene-aaa\\">hello</div>
+          <div
+            data-uid=\\"ccc aaa scene-aaa\\"
+            data-paths=\\"utopia-storyboard-uid/scene-aaa:aaa :utopia-storyboard-uid/scene-aaa\\"
```
Comment (Rheeseyb) [inline editor/src/components/canvas/ui-jsx-canvas.spec.tsx:1141]: This is missing a path, but I think that's right since the `Thing` isn't focused?

File: editor/src/components/canvas/ui-jsx-canvas.spec.tsx [hunk 13]
```diff
@@ -1222,9 +1228,14 @@ export var ${BakedInStoryboardVariableName} = (props) => {
...
             height: 400px;
           \\"
+          data-paths=\\":utopia-storyboard-uid\\"
           data-uid=\\"utopia-storyboard-uid\\"
         >
-          <div id=\\"nasty-div\\" data-uid=\\"2a7~~~1 150~~~2 2f5~~~1 aaa scene-aaa\\">
+          <div
+            id=\\"nasty-div\\"
+            data-uid=\\"2a7~~~1 150~~~2 2f5~~~1 aaa scene-aaa\\"
+            data-paths=\\":150~~~2/2a7~~~1 :150~~~2 :2f5~~~1 utopia-storyboard-uid/scene-aaa:aaa :utopia-storyboard-uid/scene-aaa\\"
```
Comment (Rheeseyb) [inline editor/src/components/canvas/ui-jsx-canvas.spec.tsx:1237]: Again, I think since `Thing` isn't focused, the paths should end at `aaa`

File: editor/src/core/performance/performance-regression-tests.spec.tsx [hunk 14]
```diff
@@ -60,8 +60,8 @@ describe('React Render Count Tests - ', () => {
     )
 
     const renderCountAfter = renderResult.getNumberOfRenders()
-    expect(renderCountAfter - renderCountBefore).toBeGreaterThan(280) // if this breaks, GREAT NEWS but update the test please :)
-    expect(renderCountAfter - renderCountBefore).toBeLessThan(290)
+    expect(renderCountAfter - renderCountBefore).toBeGreaterThan(320) // if this breaks, GREAT NEWS but update the test please :)
```
Comment (Rheeseyb) [inline editor/src/core/performance/performance-regression-tests.spec.tsx:63]: Hmmmmmmmmm.....

File: editor/src/utils/canvas-react-utils.ts [hunk 15]
```diff
@@ -128,26 +133,30 @@ function shouldIncludeDataUID(type: any): boolean {
...
     // the response was an array of elements
-    return originalResponse.map((element) => attachDataUidToRoot(element, dataUids))
+    return originalResponse.map((element) => attachDataUidToRoot(element, dataUids, paths))
   } else if (!React.isValidElement(originalResponse as any)) {
     return originalResponse
   } else {
     if (shouldIncludeDataUID(originalResponse.type)) {
       return React.cloneElement(originalResponse, {
         [UTOPIA_UIDS_KEY]: appendToUidString(originalResponse.props[UTOPIA_UIDS_KEY], dataUids),
+        [UTOPIA_PATHS_KEY]: appendToUidString(originalResponse.props[UTOPIA_PATHS_KEY], paths),
```
Comment (Rheeseyb) [inline editor/src/utils/canvas-react-utils.ts:159]: This should probably not be using `appendToUidString` (or that function should be renamed)

File: editor/src/utils/canvas-react-utils.ts [hunk 16]
```diff
@@ -269,15 +290,18 @@ const mangleExoticType = Utils.memoize(
           const originalFunction = p.children
           children = function (...params: any[]) {
             const originalResponse = originalFunction(...params)
-            return attachDataUidToRoot(originalResponse, uids)
+            return attachDataUidToRoot(originalResponse, uids, paths)
           }
         } else {
           const uidsToPass = uidsWithoutExoticUID(uids)
+          const pathsToPass = uidsWithoutExoticUID(paths)
```
Comment (Rheeseyb) [inline editor/src/utils/canvas-react-utils.ts:297]: I think this line might be wrong and might be causing some of the weirdness in the tests, so it's worth checking that this is behaving correctly with contexts and the like

File: editor/src/core/performance/performance-regression-tests.spec.tsx (hunk 14 above)
Comment (balazsbajorics) [reply editor/src/core/performance/performance-regression-tests.spec.tsx:63]: WELL, ACTUALLY this is purely because we are now running the monkey function for every test. but that is fine, as the real editor is also always running the monkey function

**Comment by balazsbajorics** [comment]
Important note: this PR does not (yet) change the selection (I didn't change the code of dom-lookup, that still uses uids), only the dom-walker's behavior! 

I'm not sure if that should be a separate PR or rolled into this one

File: editor/src/components/canvas/ui-jsx-canvas.spec.tsx (hunk 12 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/ui-jsx-canvas.spec.tsx]: this was fine, ccc is coming from an ARBITRARY_JS block

File: editor/src/components/canvas/ui-jsx-canvas-renderer/scene-root.tsx (hunk 11 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/ui-jsx-canvas-renderer/scene-root.tsx:244]: 🤔 actually I think we could. however, since the data-uid is inserted as a prop, unless we make it a not-prop, we will still print _some_ UIDs. I think in the short term it's better to leave the system of uid printing as-is, and only remove it once we believe we can fully remove it (ie not even pretend it's a prop, attach it to the parsed JSXElement in some other way)

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap (hunk 9 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:22343]: `eee/fff:aaa/bbb` was previously not in the spy data, and now it is!

Comment (balazsbajorics) [reply editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:22343]: the spy data for aaa/ddd is intact, it has been moved lower though.

File: editor/src/components/canvas/ui-jsx-canvas.spec.tsx (hunk 13 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/ui-jsx-canvas.spec.tsx:1237]: in this test case "App" is an arbitrary js. this is the last known case where the template path is going to be gibberish, but since the editor does not support editing these projects, this is not a regression, just making an existing problem explicit

File: editor/src/utils/canvas-react-utils.ts (hunk 16 above)
Comment (Rheeseyb) [reply editor/src/utils/canvas-react-utils.ts:297]: Actually ignore this :point_up: I think this should still result in it working the same way as before...

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap (hunk 8 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:21721]: I would say this is fine as is, :2a2~~~1 is an irrelevant path, and utopia-storyboard-uid/scene:BBB is valid. aaa and bbb come from arbitrary JS blocks so we can't edit them hence the missing inner template paths

File: editor/src/components/canvas/dom-walker.ts [hunk 17]
```diff
@@ -50,52 +51,20 @@ import { extractOriginalUidFromIndexedUid, getUIDsOnDomELement } from '../../cor
 import { mapDropNulls } from '../../core/shared/array-utils'
 import { optionalMap } from '../../core/shared/optional-utils'
 import { fastForEach } from '../../core/shared/utils'
+import { valid } from 'chroma-js'
```
Comment (enidemi) [inline editor/src/components/canvas/dom-walker.ts:54]: this looks like an unused import

**Review by enidemi — APPROVED** [review]

File: editor/src/components/canvas/dom-walker.ts (hunk 17 above)
Comment (balazsbajorics) [reply editor/src/components/canvas/dom-walker.ts]: indeed it was
